patches/tModLoader/Terraria.ModLoader/NPCLoader.cs

```diff
@@ -946,6 +946,84 @@ private class HookList
 			}
 		}
 
+		private static HookList HookCanGoToKingStatue = AddHook<Func<NPC, bool?>>(g => g.CanGoToKingStatue);
+
+		public static bool CanGoToKingStatue(NPC npc) {
+			bool defaultCanGo = npc.modNPC?.CanGoToKingStatue() ?? false;
+
+			switch (npc.type) {
+				case NPCID.Merchant:
+				case NPCID.Guide:
+				case NPCID.ArmsDealer:
+				case NPCID.Demolitionist:
+				case NPCID.Clothier:
+				case NPCID.GoblinTinkerer:
+				case NPCID.Wizard:
+				case NPCID.SantaClaus:
+				case NPCID.Truffle:
+				case NPCID.DyeTrader:
+				case NPCID.Painter:
+				case NPCID.WitchDoctor:
+				case NPCID.Pirate:
+				case NPCID.LightningBug: // A literal bug
+				case NPCID.Angler:
+				case NPCID.DD2Bartender:
+					defaultCanGo = true;
+					break;
+			}
+
+			foreach (GlobalNPC g in HookCanGoToKingStatue.arr) {
+				bool? canGo = g.Instance(npc).CanGoToKingStatue(npc);
+				if (canGo.HasValue) {
+					if (!canGo.Value) {
+						return false;
+					}
+					defaultCanGo = true;
+				}
+			}
+
+			return defaultCanGo;
+		}
+
+		private static HookList HookCanGoToQueenStatue = AddHook<Func<NPC, bool?>>(g => g.CanGoToQueenStatue);
+
+		public static bool CanGoToQueenStatue(NPC npc) {
+			bool defaultCanGo = npc.modNPC?.CanGoToQueenStatue() ?? false;
+
+			switch (npc.type) {
```

Comment:
What about NPCID.Sets.TeleportationImmune ? (Not 100% sure if these are related)

Reply:
That seems to be only used for teleporters in vanilla. Besides, if a modder doesn't want to let their NPC teleport to a statue at all, they can just return false in GlobalNPC.CanGoToStatue.

Reply:
Approved